### PR TITLE
fix(eslint-plugin): [no-unused-vars] SelfReferenced should see varsIgnorePattern

### PR DESCRIPTION
**Pull request by aggmoulik**
Fix: #2648
Signed-off-by: Moulik Aggarwal <qwertymoulik@gmail.com>

### CODE REVIEW

File: packages/eslint-plugin/src/rules/no-unused-vars.ts [hunk 1]
```diff
@@ -197,15 +197,20 @@ export default util.createRule<Options, MessageIds>({
...
-            messageId: 'unusedVar',
-            data: {
-              varName: variable.name,
-              action: 'defined',
-              additional: '',
-            },
-          });
+          if (
+            varsIgnorePattern &&
+            varsIgnorePattern.toString() !== variable.name
```
Comment (HolgerJeromin) [inline packages/eslint-plugin/src/rules/no-unused-vars.ts:202]: `varsIgnorePattern` is a regexp (or perhaps still a regexp string).

See usage in eslint:
`if (config.varsIgnorePattern && config.varsIgnorePattern.test(def.name.name)) {`

https://github.com/eslint/eslint/blob/551b1e9fbc463db4869e738a8228868f7323788c/lib/rules/no-unused-vars.js#L590

Comment (HolgerJeromin) [reply packages/eslint-plugin/src/rules/no-unused-vars.ts:202]: And this error should be reported also when no `varsIgnorePattern` is given.

Comment (HolgerJeromin) [reply packages/eslint-plugin/src/rules/no-unused-vars.ts:202]: Do we need a similar change in line 148?

Comment (aggmoulik) [reply packages/eslint-plugin/src/rules/no-unused-vars.ts:202]: Initially, I was confused on the same thing. 
But, I found this as my reference to go for this method:
```ts
var varIgnorePattern: "all" | "local" | {
  vars?: "all" | "local" | undefined;
  varIgnorePattern:?: string | undefined;
  args: "all" | "after-used" | "none" | undefined;
  .........
  caughtErrorsIgnorePattern?: string | undefined;
```

@bradzacher can help us in this.

Comment (aggmoulik) [reply packages/eslint-plugin/src/rules/no-unused-vars.ts:202]: Doing the regex, I am having an error
```ts
varIgnorePattern.test(variable.name)
```

Comment (aggmoulik) [reply packages/eslint-plugin/src/rules/no-unused-vars.ts:202]: Hy @HolgerJeromin, Can you guide me ?

Comment (HolgerJeromin) [reply packages/eslint-plugin/src/rules/no-unused-vars.ts:202]: No, sorry. I never hacked around around eslint

Comment (bradzacher) [reply packages/eslint-plugin/src/rules/no-unused-vars.ts:202]: So here you'll need to convert the string to a regexp:

```
const varIgnorePattern =
  typeof options === 'object' && options.varIgnorePattern != null
    ? new RegExp(options.varsIgnorePattern, "u")
    : null;
```

Comment (aggmoulik) [reply packages/eslint-plugin/src/rules/no-unused-vars.ts]: I have worked with the regex but how can I improve codecov ? @bradzacher

Comment (bradzacher) [reply packages/eslint-plugin/src/rules/no-unused-vars.ts]: you'll need some tests! right now you have none at all.

**Comment by bradzacher** [comment]
Closing in favour of #2768

**Pull request closed without merging**